core/services/ocr2/plugins/ocr2keeper/evm/offchain_lookup.go

```diff
@@ -0,0 +1,353 @@
+	if resp.StatusCode >= 400 {
+		r.setCachesOnAPIErr(upkeepId)
+	}
+
+	return body, resp.StatusCode, nil
+}
+
+// setCachesOnAPIErr when an off chain look up request fails or gets a 4xx/5xx response code we increment error count and put the upkeep in cooldown state
+func (r *EvmRegistry) setCachesOnAPIErr(upkeepId *big.Int) {
+	errCount := 1
+	cacheKey := upkeepId.String()
+	e, ok := r.apiErrCache.Get(cacheKey)
+	if ok {
+		errCount = e.(int) + 1
+	}
+
+	// With a 10m Error Cache Window every error sets the error count and resets the TTL to 10m
+	// On every error that hits during this rolling 10m the error count is increased and the cooldown period by associate is increased
+	// This means the user will suffer a max cooldown of 17m on the 10th error at which point the error cache will have expired since its window is 10m
+	// After that the user will reset to 0 and start over after a combined total of 34m in cooldown state.
+
+	// increment error count and reset expiration to shift window with last seen error
+	r.apiErrCache.Set(cacheKey, errCount, cache.DefaultExpiration)
+	// put upkeep in cooldown state for 2^errors seconds.
+	r.cooldownCache.Set(cacheKey, nil, time.Second*time.Duration(2^errCount))
+}
+
+// parseJson expects json bytes as input and offchainLookup.fields to be 'jq' style query strings for extracting data from json
+func (o *OffchainLookup) parseJson(body []byte) ([]string, error) {
+	var m map[string]any
+	err := json.Unmarshal(body, &m)
+	if err != nil {
+		return nil, err
+	}
+	results := make([]string, len(o.fields))
+	ctx, cancel := context.WithTimeout(context.Background(), 100*time.Millisecond)
+	defer cancel()
+	for i, field := range o.fields {
+		query, parseErr := gojq.Parse(field)
+		if parseErr != nil {
+			// if we can't parse the jq query use the response we are sending to the user to signal to them there was an error
+			fmt.Println(parseErr)
+			results[i] = ParseFieldError
+			continue
+		}
+
+		// always run with context
+		iter := query.RunWithContext(ctx, m)
+		for {
+			fieldValue, ok := iter.Next()
+			if !ok {
+				break
+			}
+			if fieldValueError, isError := fieldValue.(error); isError {
+				// if we have an issue with getting the value use the result response to signal that to the user
+				fmt.Println(fieldValueError)
+				results[i] = ParseFieldError
+				continue
+			}
+			if fieldValue == nil {
+				results[i] = ""
+				continue
+			}
+			if _, isMap := fieldValue.(map[string]any); isMap {
+				// if the return is a map then format it as json
+				// so the json subset can be set to the user where they can further parse if they desire
+				marshal, marshalErr := gojq.Marshal(fieldValue)
+				if marshalErr != nil {
+					// if there is an issue marshalling default to setting the field value
+					results[i] = fmt.Sprint(fieldValue)
+					continue
+				}
+				results[i] = string(marshal)
+				continue
+			}
+			results[i] = fmt.Sprint(fieldValue)
```

Comment:
We should put some explicit limits on the size of fields allowed? Gets translated to gas used in callback